File: CPP/No55.cpp
/**
 * Created by Xiaozhong on 2020/7/21.
 * Copyright (c) 2020/7/21 Xiaozhong. All rights reserved.
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    bool canJump(vector<int> &nums) {
        int max_position = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (i > max_position) return false;
            max_position = max(max_position, i + nums[i]);
        }
        return true;
    }
};

int main() {
    vector<int> nums = {2, 3, 1, 1, 4};
    Solution s;
    cout << s.canJump(nums) << endl;
}

File: Java/src/No55.java
/*
 * Copyright (c) 8/27/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No55 {
    public boolean canJump(int[] nums) {
        int maxDistance = 0;
        for (int i = 0; i < nums.length; ++i) {
            if (i <= maxDistance) {
                maxDistance = Math.max(nums[i] + i, maxDistance);
                if (maxDistance >= nums.length - 1) return true;
            }
        }
        return false;
    }
}
